docs/reference/numbers.tex

```diff
@@ -0,0 +1,601 @@
+  using type = @\seebelow@;
+};
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\ctype
+\tcode{T}'s scalar\irefiev{102-02-18}, if any.
+
+\pnum
+Each row of \tref{num.scalar} denotes a specialization.
+
+\begin{floattable}{Specializations of \tcode{number_scalar}}{num.scalar}{lll}
+\topline
+\lhdr{\fakegrammarterm{template-parameter-list}}
+                              & \chdr{\fakegrammarterm{template-argument-list}}
+                                                              & \rhdr{\Fundescx{Type}} \\ \rowsep
+\tcode{std::integral T}       & \tcode{T}                     & \tcode{T} \\ \rowsep
+\tcode{std::floating_point T} & \tcode{T}                     & \tcode{T} \\ \rowsep
+\tcode{class T}               & \tcode{std::complex<T>}       & \tcode{T} \\ \rowsep
+\tcode{class T, class U}      & \tcode{std::duration<T, U>}   & \tcode{T} \\
+\end{floattable}
+\end{itemdescr}
+
+\Sec1[num.concepts]{Concepts}
+
+\begin{itemdecl}
+template<typename T>
+concept @\deflibconcept{number}@ = is_number_v<T> && std::regular<T>;
+
+template<typename T>
+concept @\deflibconcept{ordered_number}@ = @\libconcept{number}@<T> && std::totally_ordered<T>;
+\end{itemdecl}
+
+\begin{itemdecl}
+template<class T>
+concept @\deflibconcept{number_line}@ =
+  @\libconcept{ordered_number}@<T> &&
+  requires(T& v) {
+    number_one_v<number_difference_t<T>>;
+    { ++v } -> std::same_as<T&>;
+    { --v } -> std::same_as<T&>;
+    { v++ } -> std::same_as<T>;
+    { v-- } -> std::same_as<T>;
+  };
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+Let \tcode{v} and \tcode{u} be equal objects of type \tcode{T},
+and \tcode{one} be the value \tcode{number_one_v<number_difference_t<T>>}.
+\tcode{T} models \libconcept{number_line} only if
+\begin{itemize}
+\item
+The expressions \tcode{++v} and \tcode{v++} have the same domain\irefcppx{concepts.equality}{def:expression,domain}.
+\item
+The expressions \tcode{--v} and \tcode{v--} have the same domain.
+\item
+If \tcode{v} is incrementable\irefcppx{iterator.concept.winc}{\#def:incrementable},
+then both \tcode{++v} and \tcode{v++} add \tcode{one} to \tcode{v},
+and the following expressions all equal \tcode{true}:
+\begin{itemize}
+\item \tcode{v++ == u},
+\item \tcode{((void)v++, v) == ++u}, and
+\item \tcode{std::addressof(++v) == std::addressof(v)}.
+\end{itemize}
+\item
+If \tcode{v} is decrementable\irefcppx{range.iota.view}{\#def:decrementable},
+then both \tcode{--v} and \tcode{v--} subtract \tcode{one} from \tcode{v},
+and the following expressions all equal \tcode{true}:
+\begin{itemize}
+\item \tcode{v-- == u},
+\item \tcode{((void)v--, v) == --u}, and
+\item \tcode{std::addressof(--v) == std::addressof(v)}.
+\end{itemize}
+\end{itemize}
+\end{itemdescr}
+
+\begin{itemdecl}
+template<class T, class U> concept @\defexposconceptnc{addition-with}@ =
+  @\libconcept{number}@<T> &&
+  @\libconcept{number}@<U> &&
+  requires(const T& c, const U& d) {
+    { c + d } -> std::common_with<T>;
+    { d + c } -> std::common_with<T>;
```

Comment:
I had something like <https://github.com/mpusz/mp-units/pull/492#discussion_r1334895759> at first.
I relaxed it for `addition_with<std::chrono::sys_seconds, std::chrono::seconds>`.

`vector_space<T>` will generally guarantee commutativity of addition.
That subsumes `point_space<T>`,
which subsumes `compound_point_space_for<T, number_difference_t<T>>`,
which subsumes `detail::compound_subtraction_with<T, U>`,
which subsumes `addition_with<T,U>`.

So we're working with `T` and `U = number_difference_t<T>`.
For vector spaces, those will generally be the same type.
E.g., `T = long` and `U = long`,
or `T = short` and `U = int` (thanks to integral promotion).
So `{ c + d } -> common_number_with<T>;`
and `{ d + c } -> common_number_with<T>;`
will generally do the right thing.

We're just missing `common_number_with<T, number_difference_t<T>>`.
It should suffice to add that to `vector_space`.
Then, even if `T` and `U` aren't the same type,
they are in the same family of types
(which is what I intend `common_number_with` to represent).

Reply:
> So we're working with T and U = number_difference_t<T>.
For vector spaces, those will generally be the same type.

Sure, I understand that. As I said, I consider `addition-with` as an implementation detail of `point_space`, and for this purpose, it serves well. Maybe just the name of this exposition-only concept is confusing? If it was called `point-space-addition-with` it would be clear what that serves for and people would not expect commutativity there?

Reply:
It's still commutative as defined by <https://www.electropedia.org/iev/iev.nsf/display?openform&ievref=102-01-11>.
The relaxation just makes it work with two sets (a point space with its vector space).

I'll need to consider a rename along with a semantic clarification.